src/playback/internal/playbackcontroller.cpp

```diff
@@ -215,58 +224,71 @@ void PlaybackController::togglePlay()
 
 void PlaybackController::play()
 {
-    IF_ASSERT_FAILED(playback()) {
-        return;
-    }
-
-    auto stream = playback()->midiStream();
-    sequencer()->setMIDITrack(MIDI_TRACK, stream);
-
-    RetVal<int> tick = playback()->playPositionTick();
-    if (!tick.ret) {
-        LOGE() << "unable play, err: " << tick.ret.toString();
+    IF_ASSERT_FAILED(notationPlayback() && playback()) {
         return;
     }
 
     if (m_needRewindBeforePlay) {
-        seek(tick.val);
+        seek(m_currentTick);
     } else {
         m_needRewindBeforePlay = true;
     }
 
-    sequencer()->play();
+    playback()->players()->play(m_currentSequenceId);
+
+    m_isPlaying = true;
+    m_isPlayingChanged.notify();
 }
 
 void PlaybackController::rewind(const ActionData& args)
 {
+    IF_ASSERT_FAILED(playback()) {
+        return;
+    }
+
     if (args.count() == 0) {
-        sequencer()->rewind();
+        playback()->players()->seek(m_currentSequenceId, 0);
         return;
     }
 
-    uint64_t msec = args.arg<uint64_t>(0);
-    sequencer()->seek(msec);
+    msecs_t msec = args.arg<msecs_t>(0);
+    playback()->players()->seek(m_currentSequenceId, msec);
     m_needRewindBeforePlay = false;
 }
 
-void PlaybackController::seek(int tick)
+void PlaybackController::pause()
 {
     IF_ASSERT_FAILED(playback()) {
         return;
     }
 
-    uint64_t milliseconds = secondsToMilliseconds(playback()->tickToSec(tick));
-    sequencer()->seek(milliseconds);
+    playback()->players()->pause(m_currentSequenceId);
+
+    m_isPlaying = false;
+    m_isPlayingChanged.notify();
 }
 
-void PlaybackController::pause()
+void PlaybackController::stop()
 {
-    sequencer()->pause();
+    IF_ASSERT_FAILED(playback()) {
+        return;
+    }
+
+    playback()->players()->stop(m_currentSequenceId);
+    m_isPlaying = false;
+    m_isPlayingChanged.notify();
 }
 
 void PlaybackController::resume()
 {
-    sequencer()->play();
+    IF_ASSERT_FAILED(playback()) {
+        return;
+    }
+
+    playback()->players()->play(m_currentSequenceId);
```

Comment:
I think we need some decisions to update midi data, like:
* reset buffer into midiaudiosource and request new data
* notify about new data through midistream
* explicit set new data for track here

Reply:
yes, I just realized that this is a little more complicated than I originally thought. We need to make sure that we don't accidentally re-create synthesizers / samplers because of this, and also we should only do this when we "resume" playback so that we don't do it twice on the first playback.

I will take care of this in the next PR

Reply:
I'm personally thinking about 1 and 2 options as the preferable

Reply:
turns out that it's actually not so scary, but I've found even more tiny issues during this fix :)

Anyway, I'd like to include them in another PRs, since this one is already gigantic